frontend/dockerfile/docs/reference.md

```diff
@@ -319,25 +320,133 @@ the `Dockerfile`:
 - `WORKDIR`
 - `ONBUILD` (when combined with one of the supported instructions above)
 
-Environment variable substitution will use the same value for each variable
-throughout the entire instruction. In other words, in this example:
+You can also use environment variables with `RUN`, `CMD`, and `ENTRYPOINT`
+instructions, but in those cases the variable substitution is handled by the
+command shell, not the builder. Note that instructions using the exec form
+don't invoke a command shell automatically. See [Variable
+substitution](#variable-substitution).
+
+Environment variable substitution use the same value for each variable
+throughout the entire instruction. Changing the value of a variable only takes
+effect in subsequent instructions. Consider the following example:
 
 ```dockerfile
 ENV abc=hello
 ENV abc=bye def=$abc
 ENV ghi=$abc
 ```
 
-will result in `def` having a value of `hello`, not `bye`. However,
-`ghi` will have a value of `bye` because it is not part of the same instruction
-that set `abc` to `bye`.
+- The value of `def` becomes `hello`
+- The value of `ghi` becomes `bye`
 
 ## .dockerignore file
 
 You can use `.dockerignore` file to exclude files and directories from the
 build context. For more information, see
 [.dockerignore file](https://docs.docker.com/build/building/context#dockerignore-files/).
```

Comment:
i was poking around the docs and noticed that this url is broken with the trailing slash. :) was going to open a pr, but since you're already reworking this page, do you mind making that change?

```suggestion
[.dockerignore file](https://docs.docker.com/build/building/context/#dockerignore-files).
```

Reply:
ah thanks @lizzthabet - I'll add it in here